packages/bsky/src/hydration/hydrator.ts

```diff
@@ -341,27 +343,40 @@ export class Hydrator {
     const urisLayer1 = nestedRecordUrisFromPosts(postsLayer0)
     const additionalRootUris = rootUrisFromPosts(postsLayer0) // supports computing threadgates
     const urisLayer1ByCollection = urisByCollection(urisLayer1)
-    const postUrisLayer1 = urisLayer1ByCollection.get(ids.AppBskyFeedPost) ?? []
+    const embedPostUrisLayer1 =
+      urisLayer1ByCollection.get(ids.AppBskyFeedPost) ?? []
     const postsLayer1 = await this.feed.getPosts(
-      [...postUrisLayer1, ...additionalRootUris],
+      [...embedPostUrisLayer1, ...additionalRootUris],
       ctx.includeTakedowns,
     )
     // second level embeds, ignoring any additional root uris we mixed-in to the previous layer
-    const urisLayer2 = nestedRecordUrisFromPosts(postsLayer1, postUrisLayer1)
+    const urisLayer2 = nestedRecordUrisFromPosts(
+      postsLayer1,
+      embedPostUrisLayer1,
+    )
     const urisLayer2ByCollection = urisByCollection(urisLayer2)
-    const postUrisLayer2 = urisLayer2ByCollection.get(ids.AppBskyFeedPost) ?? []
+    const embedPostUrisLayer2 =
+      urisLayer2ByCollection.get(ids.AppBskyFeedPost) ?? []
     const threadRootUris = new Set<string>()
     for (const [uri, post] of postsLayer0) {
       if (post) {
         threadRootUris.add(rootUriFromPost(post) ?? uri)
       }
     }
+    const postUrisWithThreadgates = new Set<string>()
+    for (const uri of threadRootUris) {
+      const post = postsLayer0.get(uri)
+      // post could be deleted, still want to check for existing threadgate
```

Comment:
The post wouldn't have to be deleted to be missing here—I think the main reason this method needs to fetch threadgates in this case is because there's no guarantee the caller also asked to hydrate each post's root.

The way I think we can message around this (and I appreciate the comment, definitely needed here), is to talk about checking `post.hasThreadGate` as an optimization for a common case of hydrating feed items.  When we're hydrating feed items the roots _should_ be included.  This method `hydratePosts()` doesn't rely on it, but it is allows the optimization to take place when possible.

Reply:
Oh forgive me if I have this wrong!  It looks like we _do_ fetch roots even when we're just hydrating a single post.  Curious!

Reply:
Heh yep yep, always a thing now

Reply:
How's this btw https://github.com/bluesky-social/atproto/pull/2675/commits/3aa36efc345c657aabc26502c1de6ee93993e612

Reply:
Perf 👍